dialogs/simulation/reference_density_dialog.py

```diff
@@ -0,0 +1,156 @@
+
+        uic.loadUi(gutils.get_ui_dir() / "ui_reference_density_dialog.ui", self)
+
+        self.recoil_element = recoil_element
+        self.element_simulation = element_simulation
+
+        if self.recoil_element is None:
+            QtWidgets.QMessageBox.warning(self, "Warning",
+                                                "The current simulation has "
+                                                "no recoil elements",
+                                                QtWidgets.QMessageBox.Ok,
+                                                QtWidgets.QMessageBox.Ok)
+            return
+
+        value = self.recoil_element.reference_density
+        self.scientific_spinbox = ScientificSpinBox(
+            value=value, minimum=0.01, maximum=9.99e23)
+
+        if self.recoil_element.manual_reference_density_checked:
+            self.scientific_spinbox.setDisabled(False)
+            self.userSelectionCheckBox.setChecked(True)
+        else:
+            self.scientific_spinbox.setDisabled(True)
+            self.userSelectionCheckBox.setChecked(False)
+
+        self.userSelectionCheckBox.stateChanged.connect(self._state_changed)
+
+        self.okPushButton.clicked.connect(self._accept_settings)
+        self.cancelPushButton.clicked.connect(self.close)
+
+        self.fields_are_valid = True
+
+        self.formLayout.insertRow(
+            4,
+            QtWidgets.QLabel(r"Reference density [at./cm<sup>3</sup>]:"),
+            self.scientific_spinbox)
+        self.formLayout.removeRow(self.widget)
+
+        self.isOk = False
+
+        self.exec_()
+
+    def _state_changed(self):
+        if self.userSelectionCheckBox.isChecked():
+            self.scientific_spinbox.setDisabled(False)
+        else:
+            self.scientific_spinbox.setDisabled(True)
+
+    def _density_valid(self):
+        """
+        Check if density value is valid and in limits.
+
+        Return:
+            True or False.
+        """
+        try:
+            self.scientific_spinbox.get_value()
+            return True
+        except TypeError:
+            return False
+
+    def _accept_settings(self):
+        """Function for accepting the current settings and closing the dialog
+        window.
+        """
+        if not self.fields_are_valid or not self._density_valid():
+            QtWidgets.QMessageBox.critical(
+                self, "Warning",
+                "Some of the setting values are invalid.\n"
+                "Please input values in fields indicated in red.",
+                QtWidgets.QMessageBox.Ok, QtWidgets.QMessageBox.Ok)
+            return
+
+        # If current recoil is used in a running simulation
+        if self.recoil_element is \
+                self.element_simulation.get_main_recoil():
+            if (self.element_simulation.is_simulation_running() or
+                    self.element_simulation.is_optimization_running()) and \
+                    self.name != self.recoil_element.name:
```

Comment:
There's no `self.name` anymore, so this will throw an exception if `self.element_simulation.is_optimization_running()` is True.

Also, because the automatic reference density update (#236) is simulation-wide (or at least I hope it is), then manually setting it should also apply for each `RecoilElement`. So, the checks in this `if` statement should be run for each `RecoilElement` in the simulation.

Reply:
> There's no `self.name` anymore, so this will throw an exception if `self.element_simulation.is_optimization_running()` is True.

That is not my change. What should we do with it?

Reply:
It is:
- https://github.com/JYU-IBA/potku/pull/234/commits/62f90af0ac87446eb94b17379e07549f06205dac
- https://github.com/JYU-IBA/potku/pull/234/commits/a23bb7b8c18e01d2be3422eab23726cafeb858e9

Even if it wasn't, it still needs to be fixed.

I don't know what that inequality check does in practice. Run an optimization on the **master** branch version and figure out how it works. There's probably some other way to get that name.

Reply:
Should there be own PR for this?

Reply:
> Should there be own PR for this?

It would be simpler to do the change here. After all, this is the PR that introduces the changes that can cause a crash.

Reply:
I don't know how to fix this?

Reply:
> I don't know how to fix this?

As I said, run an optimization on the master branch version and figure out how it works.

If it looks like the `and name != self.recoil_element.name:` part can be safely removed, do that. But before you do that, find out what the check actually does and why it's needed.